driver/src/main/java/org/neo4j/driver/internal/pool/PooledConnection.java

```diff
@@ -156,12 +156,15 @@ public void close()
         {
             reset( StreamCollector.NO_OP );
             sync();
-            release.accept( this );
         }
         catch (Exception ex)
         {
             dispose();
         }
+        finally
+        {
+            release.accept( this );
```

Comment:
That'll bubble up to the application layer - but this is good, if the pool rejects the connection return, that's a fatal problem that should get reported to the user. IllegalStateException happens whenever the pool state is corrupted, so this is good.